src/app/study-filters-block/study-filters-block.component.ts

```diff
@@ -7,49 +7,116 @@ import { NgbNav } from '@ng-bootstrap/ng-bootstrap';
+  public allStudyNodes: DatasetNode[] = null;
 
   @ViewChild('nav') public ngbNav: NgbNav;
 
-  public constructor(protected store: Store) {
+  public constructor(protected store: Store, public datasets: DatasetsService) {
     super(store, StudyFiltersBlockState, 'studyFilters');
   }
 
+  public showStudyFilters = false;
+
   public ngOnInit(): void {
     super.ngOnInit();
-
-    this.allStudies = new Set<string>(this.dataset.studyNames);
-
+    this.allStudies = null;
+    const datasets$ = this.datasets.getDatasetsObservable();
+    const datasetTrees = new Array<DatasetNode>();
+    datasets$.subscribe(datasets => {
+      datasets.filter(d => !d.parents.length).map(d => {
+        datasetTrees.push(new DatasetNode(d, datasets));
+      });
+      this.allStudyNodes = this.getAllStudyNodes(datasetTrees, this.dataset.id);
+      if (this.allStudyNodes && this.allStudyNodes[0]) {
+        this.allStudies = this.searchKeys(this.allStudyNodes[0]);
+      }
+    });
     this.store.selectOnce(state => state.studyFiltersBlockState as StudyFiltersBlockModel).subscribe(state => {
       // restore state
-      if (state.studyFilters.length !== 0) {
-        this.updateSelectedStudies(new Set(state.studyFilters))
-        setTimeout(() => this.ngbNav.select('studyNames'));
-      } else {
-        this.updateSelectedStudies(new Set(this.allStudies));
+      if (state['studyFilters'] !== undefined && state.studyFilters.length !== 0) {
+        const timer = setInterval(() => {
+          if (this.showStudyFilters === false) {
+            this.showStudyFilters = true;
+            this.ngbNav.select('studies');
+            this.updateSelectedStudies(new Set(state.studyFilters));
+            if (this.showStudyFilters === true) {
+              this.triggerUpdate.emit(new Set(state.studyFilters));
+              clearInterval(timer);
+            }
+          }
+        }, 200);
+      }
+    });
+  }
+
+  public searchKeys(container: DatasetNode, result: Set<string> = new Set<string>()): Set<string> {
+    container.children.forEach(containerChild => {
+      if (containerChild.children.length === 0) {
+        result.add(containerChild.dataset.id);
+      } else if (containerChild.children.length > 0) {
+        for (const child of containerChild.children) {
+          result.add(child.dataset.id);
+          this.searchKeys(child, result);
+        }
       }
     });
+    return result;
   }
 
   public onNavChange(): void {
+    if (this.ngbNav.activeId === 'studies') {
+      this.selectedStudies.add('');
+    } else {
+      this.selectedStudies = new Set<string>();
+    }
     this.store.dispatch(new SetStudyFilters([]));
     this.store.dispatch(new StateReset(StudyFiltersBlockState));
-    this.selectedStudies = new Set(this.allStudies);
+  }
+
+  private getAllStudyNodes(datasetTrees: DatasetNode[], datasetId: string): DatasetNode[] {
```

Comment:
If I understand correctly, this method searches for the dataset with id `datasetId` in the `datasetTrees` structure. The name is a bit misleading in that case - perhaps rename it to something more accurate.